# UFO Sightings Analysis

## by Justin Sierchio

In this analysis, we will be looking at reports of Unidentified Flying Objects (UFOs) from the National UFO Reporting Center. Ideally, we would like to be able to answer the following questions:

<ul>
    <li>Which areas are more likely to have sightings?</li>
    <li>What times of the year are more likely for sightings to occur?</li>
    <li>Are there clusters around military bases or other government installations?</li>
</ul>

This data is in .csv file format and is from Kaggle at: https://www.kaggle.com/NUFORC/ufo-sightings/download. More information related to the dataset can be found at: https://www.kaggle.com/NUFORC/ufo-sightings?select=complete.csv.

## Notebook Initialization

In [1]:
# Import Relevant Libraries
import pandas as pd
import numpy as np
import seaborn as sns 
import matplotlib.pyplot as plt

print('Initial libraries loaded into workspace!')

Initial libraries loaded into workspace!


In [2]:
# Upload Datasets for Study
df_UFO = pd.read_csv("scrubbed.csv");

print('Datasets uploaded!');

Datasets uploaded!


C:\Users\jmsie\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (5,9) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
# Display 1st 5 rows from UFO Sightings dataset
df_UFO.head()

datetime                  city state country     shape  \
0  10/10/1949 20:30            san marcos    tx      us  cylinder   
1  10/10/1949 21:00          lackland afb    tx     NaN     light   
2  10/10/1955 17:00  chester (uk/england)   NaN      gb    circle   
3  10/10/1956 21:00                  edna    tx      us    circle   
4  10/10/1960 20:00               kaneohe    hi      us     light   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   
1               7200              1-2 hrs   
2                 20           20 seconds   
3                 20             1/2 hour   
4                900           15 minutes   

                                            comments date posted    latitude  \
0  This event took place in early fall around 194...   4/27/2004  29.8830556   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005    29.38421   
2  Green/Orange circular disc over Chester&#44 En...   1/21/2008        53.2   
3  My older brother and twin sister were leaving ...   1/17/2004  28.9783333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004  21.4180556   

   longitude   
0  -97.941111  
1  -98.581082  
2   -2.916667  
3  -96.645833  
4 -157.803611

## Data Cleaning

In [4]:
df_UFO.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   datetime              80332 non-null  object 
 1   city                  80332 non-null  object 
 2   state                 74535 non-null  object 
 3   country               70662 non-null  object 
 4   shape                 78400 non-null  object 
 5   duration (seconds)    80332 non-null  object 
 6   duration (hours/min)  80332 non-null  object 
 7   comments              80317 non-null  object 
 8   date posted           80332 non-null  object 
 9   latitude              80332 non-null  object 
 10  longitude             80332 non-null  float64
dtypes: float64(1), object(10)
memory usage: 6.7+ MB


Given that some of the data has varying time stamps and other information, let us do a little cleaning. We'll begin with the time and data information.

In [5]:
# Clean Up Time and Date Information

# Fix datetimes
df_UFO['datetime'] = pd.to_datetime(df_UFO['datetime'], errors='coerce')
df_UFO.insert(1, 'year', df_UFO['datetime'].dt.year)
df_UFO['year'] = df_UFO['year'].fillna(0).astype(int)

# Display new 1st 5 rows
df_UFO.head(5)

datetime  year                  city state country     shape  \
0 1949-10-10 20:30:00  1949            san marcos    tx      us  cylinder   
1 1949-10-10 21:00:00  1949          lackland afb    tx     NaN     light   
2 1955-10-10 17:00:00  1955  chester (uk/england)   NaN      gb    circle   
3 1956-10-10 21:00:00  1956                  edna    tx      us    circle   
4 1960-10-10 20:00:00  1960               kaneohe    hi      us     light   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   
1               7200              1-2 hrs   
2                 20           20 seconds   
3                 20             1/2 hour   
4                900           15 minutes   

                                            comments date posted    latitude  \
0  This event took place in early fall around 194...   4/27/2004  29.8830556   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005    29.38421   
2  Green/Orange circular disc over Chester&#44 En...   1/21/2008        53.2   
3  My older brother and twin sister were leaving ...   1/17/2004  28.9783333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004  21.4180556   

   longitude   
0  -97.941111  
1  -98.581082  
2   -2.916667  
3  -96.645833  
4 -157.803611

Now let us clean up the location information (city, state, latitude, longitude).

In [6]:
# Clean up Location Information

# City information
df_UFO['city'] = df_UFO['city'].str.title() # Capitalize the first letter

# State information
df_UFO['state'] = df_UFO['state'].str.upper() # Two letter capital abbreviations

# Latitude information
df_UFO['latitude'] = pd.to_numeric(df_UFO['latitude'], errors='coerce') # Change to a numeric value

# Latitude & Longitude
df_UFO = df_UFO.rename(columns={'longitude ':'longitude'})

df_UFO.head(5)

datetime  year                  city state country     shape  \
0 1949-10-10 20:30:00  1949            San Marcos    TX      us  cylinder   
1 1949-10-10 21:00:00  1949          Lackland Afb    TX     NaN     light   
2 1955-10-10 17:00:00  1955  Chester (Uk/England)   NaN      gb    circle   
3 1956-10-10 21:00:00  1956                  Edna    TX      us    circle   
4 1960-10-10 20:00:00  1960               Kaneohe    HI      us     light   

  duration (seconds) duration (hours/min)  \
0               2700           45 minutes   
1               7200              1-2 hrs   
2                 20           20 seconds   
3                 20             1/2 hour   
4                900           15 minutes   

                                            comments date posted   latitude  \
0  This event took place in early fall around 194...   4/27/2004  29.883056   
1  1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005  29.384210   
2  Green/Orange circular disc over Chester&#44 En...   1/21/2008  53.200000   
3  My older brother and twin sister were leaving ...   1/17/2004  28.978333   
4  AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004  21.418056   

    longitude  
0  -97.941111  
1  -98.581082  
2   -2.916667  
3  -96.645833  
4 -157.803611

Given that the data is random outside the United States, we will only evaluate sightings within the US. To do this, we will create an array for all 50 US states and then only include data that is contained in one of the those 50 US states.

In [7]:
# Create array of US States
states_array = np.asarray(['AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'HI', 'IA', 'ID', 'IL', 'IN', 
                           'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ', 
                           'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VT', 'WA', 
                           'WI', 'WV', 'WY'])

In [8]:
# Only count sightings within the 50 US states

# Sort data by year
df_UFO = df_UFO[df_UFO['state'].isin(states_array)].sort_values('year')

# Make sure the latitude and longitude data only confines data to areas encompassed by the US
df_UFO = df_UFO[(df_UFO.latitude > 15) & (df_UFO.longitude < -65)]
df_UFO = df_UFO[(df_UFO.latitude > 50) & (df_UFO.longitude > -125) == False]

# Remove any other countries in the dataset
df_UFO = df_UFO[df_UFO['city'].str.contains('\(Canada\)|\(Mexico\|(UK/England)|(Australia)') == False]

C:\Users\jmsie\anaconda3\lib\site-packages\pandas\core\strings.py:1954: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


Let's once again look at the 1st 5 rows of the cleaned dataset.

In [9]:
df_UFO.head(5)

datetime  year          city state country    shape duration (seconds)  \
14868      NaT     0    Charestown    IN     NaN   circle                600   
43727      NaT     0        Malaga    NJ      us   sphere                 60   
35114      NaT     0  Port Salerno    FL      us  unknown                  2   
9117       NaT     0     Henderson    TX      us      NaN                180   
73756      NaT     0        Dennis    MA      us   circle               1260   

      duration (hours/min)                                           comments  \
14868              ten min  it&#39s was very brght :and ciricular object t...   
43727               60 sec  what i saw had no tail    it was a blue light ...   
35114   just a few seconds  two lights connected to nothing going from the...   
9117         three minutes      i saw a mothership letting off flying sausers   
73756                21:00  Bright orange orb at a very high rate of speed...   

      date posted   latitude  longitude  
14868   1/17/2004  38.453120 -85.670241  
43727   6/12/2008  39.569722 -75.048056  
35114   3/19/2002  27.143889 -80.200833  
9117    3/19/2002  32.153056 -94.799167  
73756   9/24/2012  41.735278 -70.194444

It appears our data is sufficiently cleaned to begin an analysis.

## Data Analysis

The first question we wanted to answer was were there any areas that had a higher propensity for sightings. To answer this question, let us look at the top 10 states with sightings.

In [10]:
# List the Top 10 states for UFO sightings
print('Here are the Top 10 States with UFO Sightings:\n')
print('State # Sightings (% of Total)\n')
df_UFO_TypePerc = 100. * df_UFO['state'].value_counts() / len(df_UFO);
print(df_UFO_TypePerc.head(10))

Here are the Top 10 States with UFO Sightings:

State # Sightings (% of Total)

CA    13.620765
WA     6.020251
FL     5.925633
TX     5.192696
NY     4.543079
AZ     3.794608
IL     3.732471
PA     3.642090
OH     3.420373
MI     2.923275
Name: state, dtype: float64


As we see, California leads the US with 13.6% of the continental US sightings, followed by Washington (6.0%), Florida (5.9%), and Texas (5.2%). However, these are some of the largest states. Let's normalize this data by each state's popualtion to get an idea of which states have the highest density of sightings.

Using https://state.1keydata.com/state-population-density.php, I created an Excel file with total area.

In [11]:
# Upload Population Density dataset
df_USStatesPopDensity = pd.read_excel("US_states_popdensity.xlsx");

print('Dataset uploaded!');

Dataset uploaded!


In [12]:
# Display 1st 5 rows of US States population density dataset
df_USStatesPopDensity.head(5)

rank state           name  population  land area  pop density
0     1    NJ     New Jersey     8958013       7417       1207.8
1     2    RI   Rhode Island     1056298       1045       1010.8
2     3    MA  Massachusetts     6794422       7840        866.6
3     4    CT    Connecticut     3590886       4845        741.2
4     5    MD       Maryland     6006401       9774        614.5

To accomplish our goal, we will convert the resulting number of sightings per state into a new dataframe. Then we will join it with our US population density dataset. This merged set will allow us to find which state has the highest number of sightings per populated square mile.

In [13]:
# Create new State Count Variable
value_counts = df_UFO['state'].value_counts(dropna=True, sort=True)

# Create new dataframe
df_UFO_State = pd.DataFrame(value_counts)
df_UFO_State = df_UFO_State.reset_index()
df_UFO_State.columns = ['state', 'counts']

# Display 1st 5 rows of dataset
df_UFO_State.head()

state  counts
0    CA    9645
1    WA    4263
2    FL    4196
3    TX    3677
4    NY    3217

Now that we have a new dataframe, let's join this new dataset with the population density dataset.

In [14]:
df_UFO_Density = df_UFO_State.join(df_USStatesPopDensity.set_index('state'), on='state');
df_UFO_Density.head()

state  counts  rank        name  population  land area  pop density
0    CA    9645  11.0  California  39144818.0   155959.0        251.0
1    WA    4263  23.0  Washington   7170351.0    66544.0        107.8
2    FL    4196   8.0     Florida  20271272.0    53927.0        375.9
3    TX    3677  26.0       Texas  27469114.0   261797.0        104.9
4    NY    3217   7.0    New York  19795791.0    47214.0        419.3

Now that we have our merged dataset, let's add another column for sighting density.

In [15]:
# Create column for sighting density by state
df_UFO_Density['Sighting Density (Counts/People/Square Mile)'] = df_UFO_Density['counts'] / df_UFO_Density['pop density'];
df_UFO_Density.head()

state  counts  rank        name  population  land area  pop density  \
0    CA    9645  11.0  California  39144818.0   155959.0        251.0   
1    WA    4263  23.0  Washington   7170351.0    66544.0        107.8   
2    FL    4196   8.0     Florida  20271272.0    53927.0        375.9   
3    TX    3677  26.0       Texas  27469114.0   261797.0        104.9   
4    NY    3217   7.0    New York  19795791.0    47214.0        419.3   

   Sighting Density (Counts/People/Square Mile)  
0                                     38.426295  
1                                     39.545455  
2                                     11.162543  
3                                     35.052431  
4                                      7.672311

Now we have the data necessary to find the sighting count density by state.

In [16]:
# List the Top 10 states for UFO sightings (counts / sq. mile)
print('Here are the Top 10 States with UFO Sightings/People/Square Mile:\n')
df_UFO_Density_State = df_UFO_Density.sort_values(by=['Sighting Density (Counts/People/Square Mile)'], ascending=False);
df_UFO_Density_State.head(10)

Here are the Top 10 States with UFO Sightings/People/Square Mile:



state  counts  rank        name  population  land area  pop density  \
42    AK     352  50.0      Alaska    738432.0   571951.0          1.3   
38    MT     509  48.0     Montana   1032949.0   145552.0          7.1   
27    NM     814  45.0  New Mexico   2085109.0   121356.0         17.2   
5     AZ    2687  33.0     Arizona   6828065.0   113635.0         60.1   
11    OR    1844  39.0      Oregon   4028977.0    95997.0         42.0   
1     WA    4263  23.0  Washington   7170351.0    66544.0        107.8   
0     CA    9645  11.0  California  39144818.0   155959.0        251.0   
3     TX    3677  26.0       Texas  27469114.0   261797.0        104.9   
26    NV     903  42.0      Nevada   2890845.0   109826.0         26.3   
46    WY     202  49.0     Wyoming    586107.0    97100.0          6.0   

    Sighting Density (Counts/People/Square Mile)  
42                                    270.769231  
38                                     71.690141  
27                                     47.325581  
5                                      44.708819  
11                                     43.904762  
1                                      39.545455  
0                                      38.426295  
3                                      35.052431  
26                                     34.334601  
46                                     33.666667

As one can see, Alaska leads the nation, followed by Montana, New Mexico, Arizona and Oregon. All these states are located in the western United States with clearer skies and lower population densities.

Of states with at least 100 people per square mile, Washington state leads the nation followed by California and Texas.

Let's look at the states with the lowest number of sightings per people per square mile.

In [17]:
df_UFO_Density_State.tail(10)

state  counts  rank           name  population  land area  pop density  \
44    VT     306  31.0        Vermont    626042.0     9250.0         67.7   
37    NH     534  21.0  New Hampshire   1330608.0     8968.0        148.4   
43    HI     352  13.0         Hawaii   1431603.0     6423.0        222.9   
17    MA    1355   3.0  Massachusetts   6794422.0     7840.0        866.6   
25    MD     911   5.0       Maryland   6006401.0     9774.0        614.5   
23    CT     967   4.0    Connecticut   3590886.0     4845.0        741.2   
13    NJ    1510   1.0     New Jersey   8958013.0     7417.0       1207.8   
48    DE     182   6.0       Delaware    945934.0     1954.0        484.1   
45    RI     288   2.0   Rhode Island   1056298.0     1045.0       1010.8   
50    DC      96   NaN            NaN         NaN        NaN          NaN   

    Sighting Density (Counts/People/Square Mile)  
44                                      4.519941  
37                                      3.598383  
43                                      1.579183  
17                                      1.563582  
25                                      1.482506  
23                                      1.304641  
13                                      1.250207  
48                                      0.375955  
45                                      0.284923  
50                                           NaN

As one might guess, states in the Northeast (Maryland, Connecticut, New Jersey, Delaware and Rhode Island) have the fewest sightings per people per square mile. A possible reason for this is the light pollution living in more urban environments.